cpp/src/join/conditional_join.cu

```diff
@@ -35,6 +35,103 @@
 namespace cudf {
 namespace detail {
 
+std::unique_ptr<rmm::device_uvector<size_type>> conditional_join_semi(
+  table_view const& left,
+  table_view const& right,
+  ast::expression const& binary_predicate,
+  join_kind join_type,
+  std::optional<std::size_t> output_size,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  auto right_num_rows{right.num_rows()};
+  auto left_num_rows{left.num_rows()};
+  if (right_num_rows == 0) {
+    switch (join_type) {
+      case join_kind::LEFT_ANTI_JOIN:
+        return std::make_unique<rmm::device_uvector<size_type>>(left_num_rows, stream, mr);
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr);
+      default: CUDF_FAIL("Invalid join kind."); break;
+    }
+  } else if (left_num_rows == 0) {
+    switch (join_type) {
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr);
+      default: CUDF_FAIL("Invalid join kind."); break;
+    }
+  }
+
+  auto const has_nulls = binary_predicate.may_evaluate_null(left, right, stream);
+
+  auto const parser =
+    ast::detail::expression_parser{binary_predicate, left, right, has_nulls, stream, mr};
+  CUDF_EXPECTS(parser.output_type().id() == type_id::BOOL8,
+               "The expression must produce a boolean output.");
+
+  auto left_table  = table_device_view::create(left, stream);
+  auto right_table = table_device_view::create(right, stream);
+
+  detail::grid_1d const config(left_num_rows, DEFAULT_JOIN_BLOCK_SIZE);
+  auto const shmem_size_per_block = parser.shmem_per_thread * config.num_threads_per_block;
+
+  // the code below can also be taken out in the context of semi & anti, but
+  // i will leave that for another PR for sake of being conservative
+  std::size_t join_size;
+  if (output_size.has_value()) {
```

Comment:
Like the comment says I believe that the extra overhead introduced by having more kernels for figuring out the exact size may justify it's removal because we know that the size of the join has an upper bound of the left table's size. Compromising on that extra memory allocation would lead to performance gains at runtime.

However I'll leave that for another ticket. I would also have to change the tests for conditional_joins incase this is desired.

Reply:
What are the speedups brought by the current PR?

Reply:
Using the (google based) join benchmarks, it actually does not seem to be faster. Would it make more sense to include the removal of the extra kernels for computing the join size in this PR?

The allocation of memory for the conditional left anti/semis is categorically wasteful, but this adds a new kernel at seemingly no speed up on my 3080.

Branch 24.02:
```
cudf_dev) ksm@Kashimo:~/work/cudf/cpp/build/benchmarks$ ./JOIN./JOIN_BENCH --benchmark_filter="conditional_left_semi_join|conditional_left_anti_join"
-bash: ./JOIN./JOIN_BENCH: No such file or directory
(cudf_dev) ksm@Kashimo:~/work/cudf/cpp/build/benchmarks$ ./JOIN_BENCH --benchmark_filter="conditional_left_semi_join|conditional_left_anti_join"
2024-01-04T14:21:32-05:00
Running ./JOIN_BENCH
Run on (24 X 3187.2 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x12)
  L1 Instruction 32 KiB (x12)
  L2 Unified 1280 KiB (x12)
  L3 Unified 30720 KiB (x1)
Load Average: 1.27, 10.13, 9.61
----------------------------------------------------------------------------------------------------------------------------------------------
Benchmark
                   Time             CPU   Iterations
----------------------------------------------------------------------------------------------------------------------------------------------
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit/100000/100000/manual_time               313 ms          314 ms            2
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit/100000/400000/manual_time              1123 ms         1122 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit/100000/1000000/manual_time             2750 ms         2750 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit/100000/100000/manual_time               322 ms          322 ms            2
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit/100000/400000/manual_time              1156 ms         1156 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit/100000/1000000/manual_time             2884 ms         2884 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit_nulls/100000/100000/manual_time         545 ms          544 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit_nulls/100000/400000/manual_time        1953 ms         1953 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit_nulls/100000/1000000/manual_time       4705 ms         4705 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit_nulls/100000/100000/manual_time         549 ms          549 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit_nulls/100000/400000/manual_time        2006 ms         2006 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit_nulls/100000/1000000/manual_time       4871 ms         4871 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit/100000/100000/manual_time               323 ms          323 ms            2
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit/100000/400000/manual_time              1129 ms         1129 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit/100000/1000000/manual_time             2784 ms         2784 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit/100000/100000/manual_time               321 ms          321 ms            2
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit/100000/400000/manual_time              1160 ms         1160 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit/100000/1000000/manual_time             2855 ms         2855 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit_nulls/100000/100000/manual_time         541 ms          541 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit_nulls/100000/400000/manual_time        1962 ms         1962 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit_nulls/100000/1000000/manual_time       4762 ms         4762 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit_nulls/100000/100000/manual_time         551 ms          551 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit_nulls/100000/400000/manual_time        2024 ms         2024 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit_nulls/100000/1000000/manual_time       4909 ms         4909 ms            1
```

This branch:

```
Running ./JOIN_BENCH
Run on (24 X 3187.2 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x12)
  L1 Instruction 32 KiB (x12)
  L2 Unified 1280 KiB (x12)
  L3 Unified 30720 KiB (x1)
Load Average: 1.03, 8.57, 9.10
----------------------------------------------------------------------------------------------------------------------------------------------
Benchmark
         Time             CPU   Iterations
----------------------------------------------------------------------------------------------------------------------------------------------
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit/100000/100000/manual_time               315 ms          315 ms            2
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit/100000/400000/manual_time              1150 ms         1150 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit/100000/1000000/manual_time             2798 ms         2798 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit/100000/100000/manual_time               324 ms          324 ms            2
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit/100000/400000/manual_time              1166 ms         1166 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit/100000/1000000/manual_time             2880 ms         2880 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit_nulls/100000/100000/manual_time         550 ms          550 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit_nulls/100000/400000/manual_time        1970 ms         1970 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_anti_join_32bit_nulls/100000/1000000/manual_time       4787 ms         4787 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit_nulls/100000/100000/manual_time         554 ms          554 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit_nulls/100000/400000/manual_time        2026 ms         2026 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_anti_join_64bit_nulls/100000/1000000/manual_time       4933 ms         4933 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit/100000/100000/manual_time               318 ms          318 ms            2
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit/100000/400000/manual_time              1143 ms         1143 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit/100000/1000000/manual_time             2859 ms         2859 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit/100000/100000/manual_time               324 ms          324 ms            2
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit/100000/400000/manual_time              1188 ms         1188 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit/100000/1000000/manual_time             2921 ms         2921 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit_nulls/100000/100000/manual_time         533 ms          533 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit_nulls/100000/400000/manual_time        1972 ms         1972 ms            1
ConditionalJoin<int32_t, int32_t>/conditional_left_semi_join_32bit_nulls/100000/1000000/manual_time       4850 ms         4850 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit_nulls/100000/100000/manual_time         566 ms          566 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit_nulls/100000/400000/manual_time        2069 ms         2069 ms            1
ConditionalJoin<int64_t, int64_t>/conditional_left_semi_join_64bit_nulls/100000/1000000/manual_time       4941 ms         4941 ms            1
```

Reply:
Thanks for the investigation!

Just FYI, [compare.py](https://github.com/google/benchmark/blob/main/tools/compare.py) is a handy script doing performance comparisons in this case.

The results are aligned with my expectations. We are using a memory pool thus data allocation/deallocation shouldn't be a performance bottleneck. The PR mainly removes irrelevant temporary/output data without changing the actual logic, so the performance impact wouldn't be significant. Not a requirement, but in case you are really interested in telling the difference in memory footrpint, a peak memory usage tracker can be added to the benchmark and we will definitely see the improvement, e.g. https://github.com/rapidsai/cudf/blob/fab5af24afd36a2a58fc18492bc79b4212762b96/cpp/benchmarks/io/text/multibyte_split.cpp#L196

The majority of the new function/kernel is very similar to the originals so I have a strong feeling that some small code changes together with the use of [discard_iterator](https://thrust.github.io/doc/classthrust_1_1discard__iterator.html)s should be sufficient. Let me think about this and get back to you later...

> Would it make more sense to include the removal of the extra kernels for computing the join size in this PR?

I don't have a strong preference here. Normally, keeping the PR as small as possible will facilitate the review process.

Reply:
Oh thank you. That makes the comparison a lot easier to work with lol. IIUC those are for the NVBench benchmarks right? This one is unfortunately still on the google benchmark

Reply:
The `compare.py` script linked above is from the Google Benchmarks repo. It's the other way around -- it doesn't work for NVBench. There is a separate script for NVBench comparisons provided by that repository, but I am not very familiar with it. https://github.com/NVIDIA/nvbench/blob/main/scripts/nvbench_compare.py

Reply:
MB, was referring to the peak memory usage metric, thought that was only available for NVBench(will be using personal github acc since period to contrbute to open source is now over at my company btw).

Reply:
@PointKernel I assume y'all want to prioritize other tasks for now(especially since this one will likely increase compilation times?) but was wondering if it was ok for me to continue working on this? IIUC all that is left is sinking out the useless data and updating tests right?

Reply:
Sorry for the late reply.

>  it was ok for me to continue working on this?

Definitely yes. I've just updated your base branch to `branch-24.04`. Can you please rebase your local branch with the latest  `branch-24.04` as well?

> IIUC all that is left is sinking out the useless data and updating tests right? 

Please go for it. I will take another pass for this PR in the meantime.